src/core/lib/iomgr/executor.cc

```diff
@@ -85,217 +75,260 @@ static size_t run_closures(grpc_closure_list list) {
-        gpr_atm_no_barrier_store(&g_cur_threads, cur_thread_count + 1);
-
-        g_thread_state[cur_thread_count].thd =
-            grpc_core::Thread("grpc_executor", executor_thread,
-                              &g_thread_state[cur_thread_count]);
-        g_thread_state[cur_thread_count].thd.Start();
+          static_cast<size_t>(gpr_atm_no_barrier_load(&num_threads));
+      if (cur_thread_count < max_threads) {
+        // Increment num_threads (Safe to do a no_barrier_store instead of a
+        // cas because we always increment num_threads under the
+        // 'adding_thread_lock')
+        gpr_atm_no_barrier_store(&num_threads, cur_thread_count + 1);
+
+        thd_state[cur_thread_count].thd = grpc_core::Thread(
+            name, &GrpcExecutor::ThreadMain, &thd_state[cur_thread_count]);
+        thd_state[cur_thread_count].thd.Start();
       }
-      gpr_spinlock_unlock(&g_adding_thread_lock);
+      gpr_spinlock_unlock(&adding_thread_lock);
     }
+
     if (retry_push) {
       GRPC_STATS_INC_EXECUTOR_PUSH_RETRIES();
     }
   } while (retry_push);
 }
 
-static void executor_push_short(grpc_closure* closure, grpc_error* error) {
-  executor_push(closure, error, true);
+static GrpcExecutor g_global_executor("grpc-executor");
+
+void enqueue_long(grpc_closure* closure, grpc_error* error) {
+  g_global_executor.Enqueue(closure, error, false);
 }
 
-static void executor_push_long(grpc_closure* closure, grpc_error* error) {
-  executor_push(closure, error, false);
+void enqueue_short(grpc_closure* closure, grpc_error* error) {
+  g_global_executor.Enqueue(closure, error, true);
 }
 
-static const grpc_closure_scheduler_vtable executor_vtable_short = {
-    executor_push_short, executor_push_short, "executor"};
-static grpc_closure_scheduler executor_scheduler_short = {
-    &executor_vtable_short};
+// Short-Job executor scheduler
+static const grpc_closure_scheduler_vtable global_executor_vtable_short = {
+    enqueue_short, enqueue_short, "executor-short"};
+static grpc_closure_scheduler global_scheduler_short = {
+    &global_executor_vtable_short};
+
+// Long-job executor scheduler
+static const grpc_closure_scheduler_vtable global_executor_vtable_long = {
+    enqueue_long, enqueue_long, "executor-long"};
+static grpc_closure_scheduler global_scheduler_long = {
+    &global_executor_vtable_long};
+
+void grpc_executor_init() { g_global_executor.Init(); }
 
-static const grpc_closure_scheduler_vtable executor_vtable_long = {
-    executor_push_long, executor_push_long, "executor"};
-static grpc_closure_scheduler executor_scheduler_long = {&executor_vtable_long};
+void grpc_executor_shutdown() { g_global_executor.Shutdown(); }
+
+bool grpc_executor_is_threaded() { return g_global_executor.IsThreaded(); }
+
+void grpc_executor_set_threading(bool enable) {
+  g_global_executor.SetThreading(enable);
+}
 
 grpc_closure_scheduler* grpc_executor_scheduler(
     grpc_executor_job_length length) {
```

Comment:
unrelated, but I think `grpc_executor_job_length` is a misleading name. `grpc_executor_job_type` would be better: the variable does not contain a length, but a specific category/type of job.